tensordict/nn/storage.py

```diff
@@ -220,11 +243,36 @@ def __setitem__(self, item: torch.Tensor, value: torch.Tensor) -> None:
     def __len__(self) -> int:
         return torch.sum(self.flag).item()
 
-    def contains(self, item: torch.Tensor) -> torch.Tensor:
-        index = self._to_index(item)
+    def contains(self, value: torch.Tensor) -> torch.Tensor:
+        index = self._to_index(value)
         return self.flag[index]
 
 
+class LazyFixedStorage(FixedStorage):
```

Comment:
Does it make to convert FixedStorage implementation such that by default it is lazy-initialized?

Is there any use case that we don't want lazy-initialization?

Reply:
I had the same thought, and actually that train of thought went even further... let me elaborate.

I refactored a bit FixedStorage such that we don't use modulo to get the index transform but a dict that maps the hash value to an int from 0 to M-1 where M is the max number of items we are going to store.

Also I think we don't need to rely on the Embedding class, we could just have a single tensor (akin to the weight that we're modifying) and assign values to it. That would make one less "code dependence" and clarify things IMO.

Now let's put these things together: (1) we need to convert any sparse index from 0 to N-1 with N >> M to an index from 0 to M-1 and (2) once we have that "dense" index we need to write on a storage (possibly lazily).
That looks a lot like torchrl's TensorStorage. The DynamicStorage looks a lot like torchrl's ListStorage (provided that you add the index-to-index transform explained in (1)).

So I'm starting to wonder: we are coding something with the purpose of applying it to MCTS, and recoding something very similar to what we have in torchrl (at list the second block, the Index2TensorDict part). What's missing in torchrl is
1. The TensorDict2Index (how to map a tensordict to a hash value)
2. How to map the hash value to an index from 0 to M-1
but we have the storage part, and it's pretty neat (works with prioritized values, N-dimension, arbitrary pytrees...)

Not sure what the way forward is, but here are various versions of what we could do
1. Make sure that whatever API we propose here matches TorchRL's storages one, such that we can have a minimal amount of storages coded here and just use TorchRL's when we need to
2. move things to torchrl and make an API that leverages the storages but provides more flexibility in terms of indexing. Something like:
```python
num_elts = 100
t2t_storage = TD2TDStorage(
    storage=LazyTensorStorage(num_elts), 
    in_keys=["obs", "action"], 
    out_keys=None, 
    map=TensorToIndex()) # out_keys dynamically determined
```

(1) would be a bit weird if we don't make things well bc there will be some implicit dependency on TorchRL's API that will make little sense here, but it could make sense if we have other usages of this kind of storages than MCTS (can you name any?)
(2) would make more sense if we are only thinking about MCTS

Reply:
ListStorage is pretty neat, I missed it in TorchRL. 

My typical strategy is to optimize for simplicity, in this case having the code in torchrl make many decisions more simple.  So my vote would be moving this PR to TorchRL.

At this moment, I don't have any other use case than MCTS. If future, if the same abstraction can be used to implement K-NN search, then we may have other use cases (e.g. RAG), thought I would not optimize for this potential use case.